### ISSUE TITLE
Speeding up normalized google distance

### ISSUE THREAD

**Issue by dkoslicki** [RTXteam/RTX#96]
Approaches:
1. Decorate the KG with NGD.
2. Pre-populate the cache with results for common queries.

Note the following from @finnagin:
So, I looked into it and found this under the section Frequency, Timing and Registration of E-utility URL Requests: https://www.ncbi.nlm.nih.gov/books/NBK25497/#_chapter2_Introduction_

It looks like currently 3 calls per second is the maximum allowed before they will ban your ip. However in the next section down they state that tomorrow (May 1st) we can make an account to get an API key with which the limit will be adjusted to 10 calls per second and additionally we can email to request a larger limit for our API key. Here is the paragraph from the link above:
"Coming in May 2018: API Keys

On May 1, 2018, NCBI will begin enforcing the use of API keys that will offer enhanced levels of supported access to the E-utilities. After that date, any site (IP address) posting more than 3 requests per second to the E-utilities without an API key will receive an error message. By including an API key, a site can post up to 10 requests per second by default. Higher rates are available by request (vog.hin.mln.ibcn@seitilitue). Users can obtain an API key now from the Settings page of their NCBI account (to create an account, visit www.ncbi.nlm.nih.gov/account/). After creating the key, users should include it in each E-utility request by assigning it to the new api_key parameter."

And:
I ran 15 pairwise queries and averaged about 3 seconds a query thus, it looks like pre-populating the cache for every relationship is going to be too slow. Maybe we can run queries for edges that we think will be relevant instead of all the edges in the graph? We could start by running queries for all of the edges connecting to the nodes from issue #85. (https://github.com/RTXteam/RTX/issues/85)

**Comment by edeutsch** [comment]
Sorry to blunder in here, but I thought our NGD values from PubMed was an internal dataset provided by a collaborator rather than something we call out to NCBI for? What am I missing?

**Comment by dkoslicki** [comment]
@edeutsch We get this information from NCBI eUtils via API call (returning the number of hits in abstracts mentioning the input terms). So it's not an internal dataset but something we access via [QueryNCBIeUtils.py](https://github.com/RTXteam/RTX/blob/master/code/reasoningtool/kg-construction/QueryNCBIeUtils.py).

**Comment by edeutsch** [comment]
Ah, I thought that's what Arnab's PubMed scraper was all about.. hmm. thanks for fixing me.

**Comment by dkoslicki** [comment]
Update from @finnagin:
So after deleting my local cache and running 20 different queries it took NCBI an average of ~0.378s to service one pubmed query.

I also tested running them through the whole pubmed hit count function and got an average of ~0.444s per query. Also looking at running queries through the whole normalized Google distance function it looks like queries take around the same as the pubmed hit count time. So, it looks to me like a lot of the time used in the normalized google distance calculation is coming from the queries used to convert the terms to mesh terms rather than the pubmed hit count queries themselves. I'll do some more testing and look into this and see what I might be able to do to speed things up there.

**Comment by saramsey** [comment]
@finnagin please talk to David about this

**Comment by dkoslicki** [comment]
@finnagin Since the bottleneck is converting terms to MeSH terms, how about we do the following: write a script that will convert node terms to MeSH terms for all nodes in the graph, thereby updating the cache with this information. Hopefully having populated the cache, this should remove the bottleneck. Let me know if any of this is unclear.

**Comment by finnagin** [comment]
Copying this from my and David's slack conversation:
I have been doing some more testing and found the following results:

```
-------------------------------
avg time of individual mesh conversion = 0.7887764692306518
-------------------------------
avg time of ngd w/ cached term conversion = 1.1317134857177735
-------------------------------
avg time of ngd no cache = 3.4222607135772707
-------------------------------
```

It looks like just caching the terms with roughly triple the speed of the function and since we currently have 69185 nodes in our KG caching all of terms conversions in our KG will take ~15 hours.

**Comment by dkoslicki** [comment]
@finnagin Any update on this? Is all the code in place to run on the new KG?

**Comment by finnagin** [comment]
@dkoslicki Yes, I am expanding the cache with the new nodes that were added (as of now ~94% done) but all the code is pushed and ready to work with the new KG.

**Comment by dkoslicki** [comment]
I think it will be good to move it into the directory now, that way we get a chance to test before the hackathon. You may want to check with @edeutsch to ensure that the correct directory is chosen (I don't know if the cache file location is relative or absolute, and his code may be looking for it in a different place). A simple ``find`` on the cache file name should be sufficient to find where it's currently located.

**Comment by edeutsch** [comment]
If someone can mock up a little python on how to call this thing, it would be relatively easy for me to hook it up. Yes, I think the directory matters. I think it is the case the sqlcache is looked for and goes into the CWD, but I'm not certain. I currently set up the API like this:
http://rtx.ncats.io/api/rtx/v1/ui/#!/getPubmedMeshNgd/get_pubmed_mesh_ngd
So the call is:
curl -X GET --header 'Accept: application/json' 'http://rtx.ncats.io/api/rtx/v1/getPubmedMeshNgd/malaria/hemoglobin'

Is it feasible to put the two terms at the end like that with a / separating them?

I can temporaily set the cwd to a certain place to run the command. Is the idea that there a huge sqlite cache that should be always in the same place? Note that we have several code bases so being in the "QuestionAnswering" folder is different if you're in devED vs. production, etc.

Maybe a mini 1 hr hackathon  would be needed to get this thing hooked up. In principle it's easy, but there are several little fiddly things that are unknown unknowns but I know I will stumble into while working on this and will need a consult.

**Comment by finnagin** [comment]
@dkoslicki I think it only looks in the directory that NormGoogleDistance.py is in. When you are doing a demo at the hack-a-thon will you accessing the reasoning tool code only through Eric's code or will you be accessing it directly at all?

@edeutsch I can mock up a little python on how to call it and post it in team-private sometime tomorrow though it looks like the api you linked is set up for 2 mesh terms while the cache we are talking about is for the class NormGoogleDistance which takes 2 or more pairs of ids and names from our KG, converts them to mesh terms where it can, and then finds the normalized google distance between them. (i.e. if the api calls `NormGoogleDistance.py` then it needs to take both the id and the name for all the nodes we are comparing which, if @dkoslicki needs to for the hack-a-thon, might be more than just 2) It's not that large of a sqlite cache (~1.5 GB) but it speeds up `NormGoogleDistance.py` about 3 fold so it needs to be wherever the version of `NormGoogleDistance.py` that you call is.

**Comment by finnagin** [comment]
@edeutsch I also think a mini hackathon dry run would probably be a good idea to work out the kinks and to make sure everything is where it should be.

**Comment by edeutsch** [comment]
@finnagin , okay, why don't you do a little prep work and then let's hangout and hack. Basically what I want is something like this:

from MeSHNGD import MeSHNGD
ngd = MeSHNGD()
response = ngd.getNGD(term1,term2)

where term1 and term2 are the two terms like "malaria" or "cerebral malaria"
where the returned response is a dict like this:
{
  "message": "Term 1 'malria' not found in MeSH",
  "response_code": "TermNotFound",
}
or this:
{
  "response_code": "OK",
  "value": 0.714
}

**Comment by dkoslicki** [comment]
@finnagin Minor point: if no overlap is found, this is technically not an error. So instead we should return the value of infinity (perhaps ``math.inf``)

**Comment by finnagin** [comment]
@dkoslicki Your initial implementation of normalized google distance returned `math.nan` so that is what my implementations have been returning. Should I switch them to `math.inf` instead? (If we change this we should probably double check that there are no breakages in your code down stream of calling the normalized google distance method)

Depending on which way I should implement it I'll change is so the response so that if no overlap is found we get either:
```
{
"response_code": "OK",
"value": nan
}
```
or
```
{
"response_code": "OK",
"value": inf
}
```

**Comment by edeutsch** [comment]
I'm not sure if the swagger API framework will be tolerant of those fancy numbers and might hurt users. Or JavaScript/JASON.. how about just None (null)?

**Comment by finnagin** [comment]
@edeutsch I can do that.

**Comment by finnagin** [comment]
It now returns the following when there is no overlap:

```
{
"response_code": "OK",
"value": None
}
```

**Comment by dkoslicki** [comment]
I agree: ``None`` is better

**Comment by edeutsch** [comment]
1) So, as I suspected, the cache gets put in the CWD, *not* in the location of the program. So in:
/mnt/data/orangeboard/devED/RTX/code/UI/OpenAPI/python-flask-server/
which is the CWD of the flask service, there is now
-rw-r--r-- 1 rt rt    5120 May 16 23:40 NGDCache.sqlite
a very small cache file. So the solution I will propose is that before running your program, I will switch the CWD to the kg-construction/ directory so that the big cache is used. Sound fine?

**Comment by edeutsch** [comment]
2) I don't know that there is much to be done about it at this point, but I have seen sqlite problems and multithreading. When I tried to use sqlite as a store for the feedback system, I eventually started getting some errors complaining that something was written with thread 2130830420438 but I was trying read with thread 230340340334 or some such nonsense. I don't know that we need action on this now, but I'm just saying that there is a risk that multithreaded flask may cause some problems with the sqlite caching. I haven't actually seen it yet with the MeSH caching, so maybe request.caching already deals with this. Just wanted to write this down so others can think about it.

**Comment by finnagin** [comment]
@edeutsch You're welcome! I'm get to testing it right away.

**Comment by edeutsch** [comment]
okay, thanks, it will go live tomorrow.
Are there any implications from the fact that the flask server is multithreaded and thus there may be multiple of these requests going on it parallel? If someone wrote a script that launched 10 of these at once, would there be problems?

**Comment by edeutsch** [comment]
(like with the caching or with eUtils concurrency limitations)?

**Comment by finnagin** [comment]
@edeutsch That is a good question. NCBI eUtils could be an issue. If more than 3 api calls per second come from 1 IP NCBI will block that IP. They do offer api keys that we could request that then allow 10 requests per second per api key. (thus if calls were being made from multiple IPs using the same api key in parallel they would all count towards the same limit) More info on it can be found here: https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/
This post does seem to suggest that after May 1st they may not be banning IPs anymore but rather will just display an error message if the limit is reached but, it would probably be wise to stay away from that limit just in case.

As for requests_cache I'm not sure. I found this unanswered question on SO that seems to suggest that it will spit out an error if too many threads access the same cache at the same time: https://stackoverflow.com/questions/41435295/how-to-make-requests-cache-work-with-many-concurrent-requests

**Comment by dkoslicki** [comment]
@finnagin Do you want to open a separate issue re: calls to NCBI? Or do you consider this issue resolved. Go ahead and close or (open a new issue and close).

**Comment by finnagin** [comment]
We didn't run into any issues at the hackathon with NCBI calls as far as I'm aware so I'm going to go a head and just close this issue.
